features/steps/steps.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/ONSdigital/dp-observation-importer/observation"
+	"github.com/ONSdigital/dp-observation-importer/schema"
+	"github.com/ONSdigital/dp-observation-importer/service"
+	"github.com/cucumber/godog"
+	"github.com/rdumont/assistdog"
+	"github.com/stretchr/testify/assert"
+)
+
+func (f *ImporterFeature) RegisterSteps(ctx *godog.ScenarioContext) {
+	ctx.Step(`^these observations are consumed:$`, f.theseObservationsAreConsumed)
+	ctx.Step(`^these dimensions should be inserted into the database for batch "([^"]*)":$`, f.theseDimensionsShouldBeInsertedIntoTheDatabaseForBatch)
+	ctx.Step(`^these observations should be inserted into the database for batch "([^"]*)":$`, f.theseObservationsShouldBeInsertedIntoTheDatabaseForBatch)
+	ctx.Step(`^a message stating "([^"]*)" observation\(s\) inserted for instance ID "([^"]*)" is produced$`, f.aMessageStatingObservationsInsertedForInstanceIDIsProduced)
+	ctx.Step(`^the observation batch size is set to "([^"]*)"$`, f.theObservationBatchSizeIsSetTo)
+
+	ctx.Step(`^instance "([^"]*)" on dataset-api has dimensions:$`, f.instanceOnDatasetapiHasDimensions)
+	ctx.Step(`^instance "([^"]*)" on dataset-api has headers "([^"]*)"$`, f.instanceOnDatasetapiHasHeaders)
+	ctx.Step(`^instance "([^"]*)" on dataset-api has no dimensions$`, f.instanceOnDatasetapiHasNoDimensions)
+}
+
+func (f *ImporterFeature) instanceOnDatasetapiHasDimensions(instanceID string, table *godog.Table) error {
+	dimensions, err := f.convertToDimensionsJson(table)
+	if err != nil {
+		return err
+	}
+
+	f.FakeDatasetAPI.NewHandler().Get("/instances/" + instanceID + "/dimensions").Reply(200).BodyString("{\"Items\": " + string(dimensions) + "}")
+	return nil
+}
+
+func (f *ImporterFeature) instanceOnDatasetapiHasNoDimensions(instanceID string) error {
+	f.FakeDatasetAPI.NewHandler().Get("/instances/" + instanceID + "/dimensions").Reply(200).BodyString("{\"Items\": [] }")
+	return nil
+}
+
+func (f *ImporterFeature) instanceOnDatasetapiHasHeaders(instanceID string, headers string) error {
+	type csvHeaders struct {
+		Headers []string `json:"headers"`
+	}
+	data := csvHeaders{
+		Headers: strings.Split(headers, ","),
+	}
+	b, err := json.Marshal(data)
+	if err != nil {
+		return err
+	}
+	f.FakeDatasetAPI.NewHandler().Get("/instances/" + instanceID).Reply(200).BodyString(string(b))
+	return nil
+}
+
+func (f *ImporterFeature) theseObservationsAreConsumed(table *godog.Table) error {
+
+	observationEvents, err := f.convertToObservationEvents(table)
+	if err != nil {
+		return err
+	}
+
+	signals := f.registerInterrupt()
+
+	cfg, err := config.Get()
+	if err != nil {
+		return err
+	}
+
+	// run application in separate goroutine
+	go func() {
+		_ = service.Run(context.Background(), cfg, f.serviceList, signals, "", "", "")
+	}()
+
+	// consume extracted observations
+	for _, e := range observationEvents {
+		if err := f.sendToConsumer(e); err != nil {
+			return err
+		}
+	}
+
+	time.Sleep(800 * time.Millisecond)
```

Comment:
Ideally we would use no `time.Sleep` calls in the tests.

Reply:
agreed, we didn't want to add it in but it was the only way we could find to stop the assertions running before the messages were consumed and processed by the application. We were also wary of changing how this application worked too much, and it would have needed a decent amount of change to get it to a point where we didn't need the sleep

Reply:
@CarlHembrough hope that explains it - maybe something to consider refactoring in the future but not really avoidable for now